autonet.1
.\" See LICENSE file for copyright and license details.
.Dd $Mdocdate: Feb 14 2015$
.Dt AUTONET 1
.Os
.Sh NAME
.Nm autonet
.Nd automatically connect to a known wifi network
.Sh SYNOPSIS
.Nm
.Op Fl s
.Op Fl v
.Oo
.Fl x Ar excluded_profile
.Op Fl x Ar ...
.Ar ...
.Oc
.Sh DESCRIPTION
.Nm
brings up the configured wireless interface, scans for networks,
and connects to the best match among previously configured known
profiles.
.Pp
.Nm
creates or resets a symlink from
.\" TODO: better format
.Pa /etc/hostname.<ifname>
to the best matching
.Pa /etc/hostname.d
entry (described below) it found in its scan. Then, it
execs
.Xr netstart 8
for
.Ar ifname .
.Pp
The options are as follows:
.Bl -tag -width Ds
.It Fl s
Just scan and create the symlink; don't run
.Xr netstart 8 .
This might be used to insert a call to
.Nm
prior to the systemwide
.Xr netstart 8
in
.Xr rc 8 .
.It Fl v
Print the version and exit.
.It Fl x Ar excluded_profile
Don't try to connect to the named profile, even if it shows up in
the scan.
.El
.Pp
Multiple profiles may be passed with
.Fl x
to exclude more than one profile.
.Pp
There are some things
.Nm
intentionally doesn't do:
.Bl -dash
.It
It doesn't provide any mechanism for discovery of unknown networks \(em
.Xr ifconfig 8
.Ap s scan command works just fine for that.
.It
It doesn't use any resources monitoring the connection status; it
runs once, synchronusly, and exits 0 when you're connected. If you
want to maintain a connection while you're away, it's easy enough
to write a script that calls
.Nm
when it detects that your network is down.
.El
.Sh CONFIGURATION
For each profile named in
.Nm
.Ap
s
.Pa config.h ,
there should be a
.Sm off
.Xr hostname.if 5
-formatted
.Sm on
file at
.Pa /etc/hostname.d/<ifname>.<profile> .
.Nm
creates a link to one of these files when it finds a known profile.
For instance, if
.Nm
was configured for interface iwn0 with a profile named
.Dq mynetwork ,
then the
.Dq mynetwork
.Xr hostname.if 5
configuration file would be located at
.Pa /etc/hostname.d/iwn0.mynetwork ,
and if that profile matched at runtime,
.Nm
would construct the symlink
.Pa "/etc/hostname.iwn0"
->
.Pa hostname.d/iwn0.mynetwork .
.Pp
Note that
.Nm
doesn't try to parse
.Xr hostname.if 5
files at runtime; if the profile for
.Dq mynetwork
doesn't match with the configuration in
.Pa iwn0.mynetwork
(i.e., if
.Pa config.h
specifies a different nwid or bssid from the one configured in
.Sm off
.Pa iwn0.mynetwork
),
.Sm on
.Nm
won't notice this, and will happily create a symlink to a
probably-broken network profile if the nwid in
.Pa config.h
matches. It is your responsibility to ensure that your
.Pa config.h
lines up with your
.Pa hostname.d .
.Pp
.Nm
doesn't clear any interface configuration aside from BSSID. Therefore,
to switch from a secured network to an open network, you'll need
an explicit "-wpakey" in the
.Xr hostname.if 5
file for the open network.
.Sh FILES
.Bl -tag -width "/etc/hostname.d/XXX.YYYYYY   "
.It Pa /etc/hostname.d/XXX.YYYYYY
Interface- and network-specific profiles.
.El
.Sh EXIT STATUS
.Nm
exits 0 on success, 1 if any of its calls failed, and 2 if it found
no networks that matched a known profile.
.Sh SEE ALSO
.Xr hostname.if 5 ,
.Xr ifconfig 8 ,
.Xr netstart 8 ,
.Xr rc 8
.Sh AUTHORS
.An Steven Dee
.Ao
.Mt steve@smartercode.net
.Ac


autonet.c
/* See LICENSE file for copyright and license details. */
#include <sys/types.h>
#include <sys/socket.h>
#include <sys/ioctl.h>
#include <sys/stat.h>
#include <sys/queue.h>

#include <net/if.h>
#include <net80211/ieee80211.h>
#include <net80211/ieee80211_ioctl.h>

#include <assert.h>
#include <err.h>
#include <errno.h>
#include <stdbool.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>

#include "arg.h"

/*
 * Network profile preference.
 *
 * Maps discovered network nwids and/or bssids to known profile names.
 * Matches against nwid if non-NULL, and against bssid if other than BSSID_ANY.
 * If filename is NULL, nwid is the profile name.
 */
typedef struct {
	char*   nwid;
	uint8_t bssid[IEEE80211_ADDR_LEN];
	char*   filename;
} NetPref;

#define BSSID_ANY "\xff\xff\xff\xff\xff\xff"

#include "config.h"

#define HOSTNAME_IF "/etc/hostname." IFNAME

static char* const connect_cmd[] =
	{ "/bin/sh", "/etc/netstart", IFNAME, NULL };
static char* const connect_env[] = { NULL };
char* argv0;
static bool sflag;


static void
usage(void)
{
	fprintf(stderr,
	        "usage: %s [-s] [-v] [[-x excluded_profile] [-x ...] ...]\n"
	        "automatic wifi network chooser.\n"
	        "Creates an appropriate /etc/hostname."IFNAME" and "
	        "execs /etc/netstart "IFNAME".\n",
	        argv0);
	exit(1);
}

static const char*
NetPref_profile(const NetPref* np)
{
	return np->filename ? np->filename : np->nwid;
}

static bool
NetPref_match(const NetPref* np, const struct ieee80211_nodereq* nr)
{
	bool bssids_match, nwids_match;

	bssids_match =
		!memcmp(np->bssid, BSSID_ANY, IEEE80211_ADDR_LEN) ||
		!memcmp(np->bssid, nr->nr_bssid, IEEE80211_ADDR_LEN);
	nwids_match =
		np->nwid && strlen(np->nwid) == nr->nr_nwid_len &&
		!strncmp(np->nwid, (const char*)nr->nr_nwid, nr->nr_nwid_len);
	return bssids_match && nwids_match;
}

/*
 * Set up the hostname.if(5) symlink and exec netstart(8).
 *
 * Prints the profile name.
 * Links /etc/hostname.<ifname> -> /etc/hostname.d/<ifname>.<profile>.
 * Dies if /etc/hostname.<ifname> exists and is not a symlink.
 */
static void
NetPref_connect(const NetPref* np)
{
	struct stat sb;
	int r;
	char buf[256];
	const char *profile;

	profile = NetPref_profile(np);
	assert(profile);
	printf("network %s\n", profile);

	/* replace hostname.if(5) symlink for ifname */
	if (lstat(HOSTNAME_IF, &sb) == -1) {
		if (errno != ENOENT)
			err(1, "lstat");
	}
	else {
		if (!(S_ISLNK(sb.st_mode)))
			errx(1, HOSTNAME_IF " is not a symlink");
		if (unlink(HOSTNAME_IF) == -1)
			err(1, "unlink");
	}
	r = snprintf(buf, sizeof(buf), "hostname.d/"IFNAME".%s", profile);
	if (r < 0)
		err(1, "snprintf");
	assert((size_t)r < sizeof(buf));
	if (symlink(buf, HOSTNAME_IF) == -1)
		err(1, "symlink");
	if (sflag) {
		/* just symlink and exit */
		exit(0);
	}
	else {
		/* exec netstart(8) */
		(void) execve(connect_cmd[0], connect_cmd, connect_env);
		err(1, "execve");
	}
}

int
main(int argc, char* argv[])
{
	struct ieee80211_nodereq_all na;
	struct ieee80211_nodereq nr[512];
	struct ieee80211_bssid bssid;
	struct ifreq ifr;
	int s, i;
	const NetPref *np;
	static SLIST_HEAD(np_excludes, np_exclude) excludes =
		SLIST_HEAD_INITIALIZER(excludes);
	struct np_exclude {
		char* npe_name;
		SLIST_ENTRY(np_exclude) npe_next;
	} *npe;

	ARGBEGIN {
	case 's':
		if (sflag)
			usage();
		sflag = true;
		break;
	case 'v':
		fputs("autonet-"VERSION
		      ", (c) 2015 Steven Dee, see LICENSE for details\n",
		      stderr);
		exit(1);
	case 'x':
		npe = malloc(sizeof *npe);
		npe->npe_name = EARGF(usage());
		SLIST_INSERT_HEAD(&excludes, npe, npe_next);
		break;
	default:
		usage();
	} ARGEND;

	s = socket(AF_INET, SOCK_DGRAM, 0);
	if (s == -1)
		err(1, "socket");
	/* TODO: exit successfully if network already active */

	/* Unset any previously configured BSSID */
	bzero(&bssid.i_bssid, sizeof(bssid.i_bssid));
	(void) strlcpy(bssid.i_name, IFNAME, sizeof(bssid.i_name));
	if (ioctl(s, SIOCS80211BSSID, &bssid) == -1)
		warn("SIOCS80211BSSID");

	/* Bring the interface up */
	bzero(&ifr, sizeof(ifr));
	(void) strlcpy(ifr.ifr_name, IFNAME, sizeof(ifr.ifr_name));
	if (ioctl(s, SIOCGIFFLAGS, (caddr_t)&ifr) == -1)
		err(1, "SIOCGIFFLAGS");
	if (!(ifr.ifr_flags & IFF_UP)) {
		ifr.ifr_flags |= IFF_UP;
		if (ioctl(s, SIOCSIFFLAGS, (caddr_t)&ifr) == -1)
			err(1, "SIOCSIFFLAGS");
	}

	/* Scan for networks */
	if (ioctl(s, SIOCS80211SCAN, (caddr_t)&ifr) == -1)
		err(1, "SIOCS80211SCAN");
	bzero(&na, sizeof(na));
	bzero(&nr, sizeof(nr));
	na.na_node = nr;
	na.na_size = sizeof(nr);
	(void) strlcpy(na.na_ifname, IFNAME, sizeof(na.na_ifname));
	if (ioctl(s, SIOCG80211ALLNODES, &na) == -1)
		err(1, "SIOCG80211ALLNODES");

	if (close(s) == -1)
		warn("close");

	/* Search for a match in order of preference */
	for (np = &networks[0]; NetPref_profile(np); np++) {
		SLIST_FOREACH(npe, &excludes, npe_next)
			if(!strcmp(npe->npe_name, NetPref_profile(np)))
				goto continue_outer;
		for (i = 0; i < na.na_nodes; i++)
			if (NetPref_match(np, &nr[i]))
				NetPref_connect(np);
continue_outer: ;
	}
	errx(2, "no known network found");
}
